DOCSP-35951: write operations delete docs (#2829)
jenssegers/laravel-mongodb: docs/fundamentals/write-operations.txt

      :end-before: end model array positional

   .. output::
      :language: json
      :visible: false

      {
        "_id": "660e...",
        "performer": "Mitsuko Uchida",
        "genres": [
          "classical",
          "contemporary"
        ],
        "updated_at": ...,
        "created_at": ...
      }

To learn more about array update operators, see :manual:`Array Update Operators </reference/operator/update-array/>`
in the {+server-docs-name+}.

.. _laravel-fundamentals-delete-documents:

Delete Documents
----------------

In this section, you can learn how to delete documents from a MongoDB collection
by using {+odm-short+}. Use delete operations to remove data from your MongoDB
database permanently.

This section provides examples of the following delete operations:

- :ref:`Delete one document <laravel-fundamentals-delete-one>`
- :ref:`Delete multiple documents <laravel-fundamentals-delete-many>`

To learn about the Laravel features available in {+odm-short+} that modify
delete behavior, see the following sections:

- :ref:`Soft delete <laravel-model-soft-delete>`, which lets you mark
  documents as deleted instead of removing them from the database

> - :ref:`Pruning <laravel-model-pruning>`, which lets you define conditions
>   in which a document should be deleted automatically

Reviewer: ```suggestion
- :ref:`Pruning <laravel-model-pruning>`, which lets you define conditions
  that mark documents for automatic deletion
```

Author: Thanks for pointing out the unclear wording. Since the conditions don't mark documents directly, so maybe I'll use "define conditions that qualify a document for automatic deletion".

.. _laravel-fundamentals-delete-one:

Delete a Document Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can delete one document in the following ways:

> - Call the  ``destroy()`` method on the model to delete a document by the
>   primary key.

Reviewer: ```suggestion
- Call the  ``destroy()`` method on the model to delete a document by matching the
  primary key.
```

Author: I think the existing sentence and suggestion might be omitting a detail necessary to understand what to pass it, so I'll figure out how to reword. Will also update the description in the corresponding section for deleting multiple docs.

- Call the ``delete()`` method on an instance of the model
- Chain methods to retrieve and delete an instance of a model by calling the
  ``delete()`` method.

The following example shows how to delete a document by passing the value of
its primary key to the ``destroy()`` method:

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin model delete by id
   :end-before: end model delete by id

When the ``destroy()`` method succeeds, it returns the number of documents
deleted.

If the primary key value does not match any documents, the ``destroy()`` method
returns returns  ``0``.

The following example shows how to delete a document by calling ``delete()``

> on an instance of the model:

Reviewer: S: separate the three examples into subsections or move into bullet points, but as of now stacking them is a bit confusing

Author: I agree with this and had considered earlier on, but wanted to keep all examples at the same heading level. Both Insert and Update use the same formatting with the same intention.

I think once these sections (Insert, Modify, Delete) are split into individual pages similar to drivers docs are separate pages,  these can be their own sections.

In the meantime, I'll add literalinclude captions (including examples in the prior sections) to identify which method is being demonstrated.

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin delete one model
   :end-before: end delete one model

When the ``delete()`` method succeeds, the operation returns the number of
documents deleted.

If the retrieve part of the call does not match any documents in the collection,
the operation returns ``0``.

The following example shows how to chain calls to retrieve the first
matching document and delete it:

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin model delete one fluent
   :end-before: end model delete one fluent

.. include:: /includes/fact-orderby-id.rst

When the ``delete()`` method succeeds, it returns the number of documents
deleted.

If the ``where()`` method does not match any documents, the ``delete()`` method
returns returns  ``0``.

.. _laravel-fundamentals-delete-many:

Delete Multiple Documents Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can perform deletes on multiple documents in the following ways:

Reviewer: ```suggestion
You can delete multiple documents in the following ways:
```

in the {+server-docs-name+}.

.. _laravel-fundamentals-delete-documents:

Delete Documents
----------------

In this section, you can learn how to delete documents from a MongoDB collection
by using {+odm-short+}. Use delete operations to remove data from your MongoDB
database permanently.

This section provides examples of the following delete operations:

- :ref:`Delete one document <laravel-fundamentals-delete-one>`
- :ref:`Delete multiple documents <laravel-fundamentals-delete-many>`

To learn about the Laravel features available in {+odm-short+} that modify
delete behavior, see the following sections:

- :ref:`Soft delete <laravel-model-soft-delete>`, which lets you mark
  documents as deleted instead of removing them from the database
- :ref:`Pruning <laravel-model-pruning>`, which lets you define conditions
  that qualify a document for automatic deletion

.. _laravel-fundamentals-delete-one:

Delete a Document Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can delete one document in the following ways:

- Call the ``destroy()`` method on the model and pass it a primary key to

>   delete the matching document.

Reviewer: S: revise to avoid the phrasing "pass it"
Ie
```suggestion
- Call the ``destroy()`` method on the model while passing a primary key to
  delete the matching document.
```

Author: Discussed on Slack. Replacing with a adverbial phrase to describe the first part of the sentence.

   :end-before: end delete one model
   :caption: Delete the document by calling the delete() method on an instance.

When the ``delete()`` method succeeds, the operation returns the number of
documents deleted.

If the retrieve part of the call does not match any documents in the collection,
the operation returns ``0``.

The following example shows how to chain calls to retrieve the first
matching document and delete it:

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin model delete one fluent
   :end-before: end model delete one fluent
   :caption: Delete the matching document by chaining the delete() method.

.. include:: /includes/fact-orderby-id.rst

When the ``delete()`` method succeeds, it returns the number of documents
deleted.

If the ``where()`` method does not match any documents, the ``delete()`` method
returns returns ``0``.

.. _laravel-fundamentals-delete-many:

Delete Multiple Documents Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can delete multiple documents in the following ways:

- Call the ``destroy()`` method on the model and pass it multiple primary

>   keys to delete the matching documents.

Reviewer: S: same comment as above re: pass it

To learn more about array update operators, see :manual:`Array Update Operators </reference/operator/update-array/>`
in the {+server-docs-name+}.

.. _laravel-fundamentals-delete-documents:

Delete Documents
----------------

In this section, you can learn how to delete documents from a MongoDB collection
by using {+odm-short+}. Use delete operations to remove data from your MongoDB
database permanently.

This section provides examples of the following delete operations:

- :ref:`Delete one document <laravel-fundamentals-delete-one>`
- :ref:`Delete multiple documents <laravel-fundamentals-delete-many>`

To learn about the Laravel features available in {+odm-short+} that modify
delete behavior, see the following sections:

- :ref:`Soft delete <laravel-model-soft-delete>`, which lets you mark
  documents as deleted instead of removing them from the database
- :ref:`Pruning <laravel-model-pruning>`, which lets you define conditions
  that qualify a document for automatic deletion

.. _laravel-fundamentals-delete-one:

Delete a Document Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can delete one document in the following ways:

> - Call the ``destroy()`` method on the model, passing it the primary key of
>   the document to be deleted.
> - Call the ``delete()`` method on an instance of the model.
> - Chain methods to retrieve and delete an instance of a model by calling the
>   ``delete()`` method.

Reviewer 2: `$model->delete()` should be the first in the list as it's probably the more used.

The alternate `Model::destroy($ids)` accepts a single id, a list of ids, or a list of model intances.

```suggestion
- Call the ``$model->delete()`` method on an instance of the model.
- Call the ``Model::destroy($ids)`` method on the model class, passing it the id of
  the document to be deleted.
- Chain methods to retrieve and delete an instance of a model by calling the
  ``delete()`` method.
```

Author: Question:
Is the "$model->delete()" a common way to represent a placeholder for an Eloquent model for a Laravel developer audience? 

Otherwise, what would be a common way to explain calling on a method on an instance of a model? I could be wrong, but I suspect the current suggestion might be repetitive (the instance and method are shown, but it's still described as a method).

Reviewer 2: I believe that code is more explicit than words. This is the convention I use systematically on this project when I want to emphasize the scope of a method.

I've found examples of `Model::method` in [Eloquent doc](https://laravel.com/docs/11.x/eloquent), but not `$model->method`. See https://laravel.com/docs/11.x/eloquent#deleting-models

Author: I'm fine with changing conventions, but is this something we could come back to and decide later? I think it might involve updating several of the currently published docs to make them use it consistently. If so, I can ticket that.

   :caption: Delete the document by calling the delete() method on an instance.

When the ``delete()`` method succeeds, the operation returns the number of
documents deleted.

If the retrieve part of the call does not match any documents in the collection,
the operation returns ``0``.

The following example shows how to chain calls to retrieve the first
matching document and delete it:

.. literalinclude:: /includes/fundamentals/write-operations/WriteOperationsTest.php
   :language: php
   :dedent:
   :start-after: begin model delete one fluent
   :end-before: end model delete one fluent
   :caption: Delete the matching document by chaining the delete() method.

.. include:: /includes/fact-orderby-id.rst

When the ``delete()`` method succeeds, it returns the number of documents
deleted.

If the ``where()`` method does not match any documents, the ``delete()`` method
returns returns ``0``.

.. _laravel-fundamentals-delete-many:

Delete Multiple Documents Examples
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can delete multiple documents in the following ways:

- Call the ``destroy()`` method on the model, passing the primary keys of
  the documents to be deleted.

>   keys to delete the matching documents.

Reviewer 2: This sentence seems incomplete. `Model::destroy($ids)` accepts a list of ids, or a list of model instances.

Note that using this method with a large number of ids is unefficient as it sends a distinct command for each id. For performance it's better using `Model::whereIn('id', $ids)->delete()`

Author: Good catch. I'll add an admonition where the example is currently placed to mention the performance issue.

To learn more about array update operators, see :manual:`Array Update Operators </reference/operator/update-array/>`
in the {+server-docs-name+}.

.. _laravel-fundamentals-delete-documents:

Delete Documents
----------------

In this section, you can learn how to delete documents from a MongoDB collection
by using {+odm-short+}. Use delete operations to remove data from your MongoDB

> database permanently.

Reviewer 2: "permanently" is wrong when the `SoftDeletes` trait is used. But you mention it just below, so maybe that's okay.

Author: Thanks for providing that perspective, I'll update the text to make that more clear!
